### ISSUE TITLE
IJW tests continue failing on Windows 7

### ISSUE THREAD

**Issue by echesakovMSFT** [dotnet/coreclr#21796]
I see these failures in my private runs in Azure DevOps (Checked build) as well as in buildpipeline (Release build) on Windows 7

**An example of failures:**

**Checked** (e.g. https://mc.dot.net/#/user/4ae9b87d-f80b-4977-8984-b4c67c7cd398/pr~2Fdotnet~2Fcoreclr~2Frefs~2Fpull~2F2~2Fmerge/test~2Ffunctional~2Fcli~2Fpri0/20190102.71/workItem/Interop.IJW/wilogs)
```
Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'IjwNativeDll, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.
         at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, IntPtr ptrLoadContextBinder)
         at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, IntPtr ptrLoadContextBinder)
         at System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, StackCrawlMark& stackMark, IntPtr pPrivHostBinder)
         at System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, StackCrawlMark& stackMark)
         at System.Reflection.Assembly.Load(String assemblyString)
         at ManagedCallingNative.ManagedCallingNative.Main(String[] args)

Return code:      1
Raw output file:      C:\dotnetbuild\work\d0bbea5e-181f-468b-ad47-e70059b15b5c\Work\6499e7bf-8622-4ac9-abcc-ac2753c5b9c5\Unzip\Reports\Interop.IJW\ManagedCallingNative\ManagedCallingNative\ManagedCallingNative.output.txt
Raw output:
BEGIN EXECUTION
       "C:\dotnetbuild\work\d0bbea5e-181f-468b-ad47-e70059b15b5c\Payload\corerun.exe" ManagedCallingNative.exe 
      Beginning scenario: Calling from managed to native IJW code
      Beginning test case Load IJW dll via reflection at 1/3/2019 3:35:08 AM
      Random seed: 20010415; set environment variable CORECLR_SEED to this value to repro
     

      Expected: 100
      Actual: -532462766
      END EXECUTION - FAILED
      FAILED
      Test Harness Exitcode is : 1
      
To run the test:
> set CORE_ROOT=C:\dotnetbuild\work\d0bbea5e-181f-468b-ad47-e70059b15b5c\Payload
> C:\dotnetbuild\work\d0bbea5e-181f-468b-ad47-e70059b15b5c\Work\6499e7bf-8622-4ac9-abcc-ac2753c5b9c5\Unzip\ManagedCallingNative\ManagedCallingNative\ManagedCallingNative.cmd

      Expected: True
      Actual:   False
      Stack Trace:
       F:\vsagent\78\s\bin\tests\Windows_NT.x64.Checked\TestWrappers\Interop.IJW\Interop.IJW.XUnitWrapper.cs(109,0): at Interop_IJW._ManagedCallingNative_ManagedCallingNative_ManagedCallingNative_._ManagedCallingNative_ManagedCallingNative_ManagedCallingNative_cmd()
```

**Release** (e.g. https://mc.dot.net/#/product/netcore/30/source/official~2Fcoreclr~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20190103.03/workItem/Interop.IJW.XUnitWrapper/analysis/xunit/Interop_IJW._ManagedCallingNative_ManagedCallingNative_ManagedCallingNative_~2F_ManagedCallingNative_ManagedCallingNative_ManagedCallingNative_cmd) 
```
Unhandled Exception: System.BadImageFormatException: Could not load file or assembly 'IjwNativeDll, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'. An attempt was made to load a program with an incorrect format. ---> System.BadImageFormatException: An attempt was made to load a program with an incorrect format. (Exception from HRESULT: 0x8007000B)
   --- End of inner exception stack trace ---
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, IntPtr ptrLoadContextBinder)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, IntPtr ptrLoadContextBinder)
   at System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, StackCrawlMark& stackMark, IntPtr pPrivHostBinder)
   at System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.Load(String assemblyString)
   at ManagedCallingNative.ManagedCallingNative.Main(String[] args)


Return code:      1
Raw output file:      C:\\dotnetbuild\\work\\111b4f1d-ef18-4dee-b5f0-1be2df78d6d0\\Work\\c1be48e3-db4f-40c3-bf1f-f264e42c2aaa\\Unzip\\Reports\\Interop.IJW\\ManagedCallingNative\\ManagedCallingNative\\ManagedCallingNative.output.txt
Raw output:
BEGIN EXECUTION
 \"C:\\dotnetbuild\\work\\111b4f1d-ef18-4dee-b5f0-1be2df78d6d0\\Payload\\corerun.exe\" ManagedCallingNative.exe 
Beginning scenario: Calling from managed to native IJW code
Beginning test case Load IJW dll via reflection at 1/3/2019 3:07:54 PM
Random seed: 20010415; set environment variable CORECLR_SEED to this value to repro

Expected: 100
Actual: -532462766
END EXECUTION - FAILED
FAILED
Test Harness Exitcode is : 1
```

It seems that the tests fail to load *api-ms-win-core-** DLLs while searching in test directory (i.e. ManagedCallingNative) and %PATH% (but not in the directory containing corerun.exe - is this desired behavior?). 

After I had installed Windows 10 Universal C Runtime for Windows 7 from https://www.microsoft.com/en-us/download/details.aspx?id=48234 the original issue disappeared and now the tests fail with the assertion

```
BEGIN EXECUTION
 "C:\git\Core_Root\corerun.exe" ManagedCallingNative.exe 
Beginning scenario: Calling from managed to native IJW code
Beginning test case Load IJW dll via reflection at 1/3/2019 4:44:04 PM
Random seed: 20010415; set environment variable CORECLR_SEED to this value to repro


Assert failure(PID 2700 [0x00000a8c], Thread: 2688 [0x0a80]): TypeFromToken(tok) == mdtMethodDef || TypeFromToken(tok) == mdtMemberRef

<no module>! <no symbol> + 0x0 (0x00000000`00000000)
    File: d:\git\coreclr2\src\vm\ceeload.cpp Line: 6566
    Image: C:\git\Core_Root\CoreRun.exe

Expected: 100
Actual: -1073740286
END EXECUTION - FAILED
FAILED
```
The same effect could be done by adding Core_Root to the Path.

At this point, I am not sure how to proceed.

@morganbr @AaronRobinsonMSFT Can you please take a look at the failure? I am planning to disable the IJW tests for now to get green CI - please let me know if you object

**Comment by morganbr** [comment]
@echesakovMSFT , your commit doesn't seem to have touched this test. Is it disabled somewhere else?

**Comment by echesakovMSFT** [comment]
I haven't disabled the tests - instead I excluded Windows.7.Amd64 from running against PR in AzDO in #21926

**Comment by AaronRobinsonMSFT** [comment]
cc @jkoritzinsky

**Comment by RussKeldorph** [comment]
_NativeCallingManaged_NativeCallingManaged_NativeCallingManaged_cmd is failing similarly

**Comment by jkoritzinsky** [comment]
The first issue (the BadImageFormatException) looks like its happening because `IjwNativeDll.dll` was linked to `api-ms-win-crt-runtime-l1-1-0.dll`. Since the Win32 API contract assemblies don't exist on Win7, that would explain the load failure. I'm still trying to get a repro environment with a debug build so I can determine what's causing the assert.

**Comment by echesakovMSFT** [comment]
@jkoritzinsky Yes you are right - try installing Windows 10 Universal C Runtime for Windows 7 from https://www.microsoft.com/en-us/download/details.aspx?id=48234

**Comment by jkoritzinsky** [comment]
I finally got a repro under a debugger with a debug runtime and release-mode test. I'll see if I can figure out what's going on.

**Comment by jkoritzinsky** [comment]
I've figured out the problem. On Windows 7, the desktop version of `mscoree.dll` is being loaded by something (which I'm still trying to track down, looks like it's getting loaded when `System.Private.CoreLib` is first loaded) before we can load our fake `mscoree.dll` shim into the process. The desktop `mscoree.dll` goes in and patches the vtfixup table in the IJW image, which then causes coreclr to fail loading it since the tokens have already been replaced with delayed-activation thunks.

Once we finish adding IJW support (#22636 and dotnet/core-setup#5185) and have an updated MSVC++ compiler that can produce IJW assemblies targeting .NET Core, then this problem will no longer be an issue since the IJW loading process will only touch a shim of the new ijwhost and won't touch `mscoree.dll` at all.

My suggestion would be to disable these tests on Windows 7 if possible until we have the new IJW activation support and compiler.

cc: @AaronRobinsonMSFT @jeffschwMSFT
